addons/binding/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoStatusProvider.java

```diff
@@ -0,0 +1,723 @@
+                connTivoConnectRetry(false);
+                errCount = 0;
+
+            }
+
+            TivoStatusData myTivo = statusGet();
+
+            if (myTivo != null) {
+                logger.debug(" statusRefresh '{}' - status data FOUND in CURRENT CHECK - '{}'",
+                        tivoConfigData.getCfgIdentifier(), myTivo.toString());
+
+            } else {
+                logger.debug(" statusRefresh '{}' - status data NOT FOUND in CURRENT CHECK!  Attempt '{}'.",
+                        tivoConfigData.getCfgIdentifier(), errCount + 1);
+            }
+
+            if (myTivo != null && myTivo.isCmdOk()) {
+                // We found valid TiVo status values, updating
+                tivoStatusData = myTivo;
+                tivoHandler.updateTivoStatus(tivoStatusData);
+            } else {
+                errCount++;
+                logger.debug(" statusRefresh '{}' - returned no new status data, the existing status will be retained",
+                        tivoConfigData.getCfgIdentifier());
+            }
+
+            if (myTivo != null && tivoStatusData != null && tivoStatusData.getMsg().equals("NO_STATUS_DATA_RETURNED")
+                    && myTivo.getMsg().equals("NO_STATUS_DATA_RETURNED")) {
+                // Force a disconnect so we can get some decent status on next poll
+                connTivoConnectRetry(false);
+            }
+            return tivoStatusData;
+        }
+        logger.debug(" statusRefresh '{}' - status polling actions are PAUSED/SKIPPED, while we execute a command",
+                tivoConfigData.getCfgIdentifier());
+        return tivoStatusData;
+    }
+
+    /**
+     * {@link cmdTivoCommand} sends a command to the Tivo.
+     *
+     * @param pCmd the command string from the item.
+     * @return {@link TivoStatusData} object.
+     */
+    public TivoStatusData cmdTivoSend(String pCmd) {
+
+        boolean oC = false; // Tracks if this step caused a connection to be opened
+        TivoStatusData tmpStatus = null;
+        int retryCount = 2;
+
+        pausePolling = true;
+
+        // Open a connection if one does not already exist
+        if (!connTestSocket()) {
+            connTivoConnectRetry(true);
+            oC = true;
+        }
+
+        tmpStatus = cmdTivoSendRetry(pCmd, retryCount);
+
+        // Will close the connection after operation if we are not keeping one open.
+        if (!tivoConfigData.isCfgKeepConnOpen() && oC && !tivoStatusData.getChScan()) {
+            connTivoConnectRetry(false);
+        }
+        pausePolling = false;
+        return tmpStatus;
+
+    }
+
+    /**
+     * {@link cmdTivoSendRetry} sets the TiVo status, executes the specified number of retries.
+     *
+     * @param pCmd the command string from the item.
+     * @param retryCount the retry count.
+     * @return {@link TivoStatusData} object.
+     */
+    private TivoStatusData cmdTivoSendRetry(String pCmd, int retryCount) {
```

Comment:
most of the calls to this method happen from `TivoHandler#handleCommand` and the method might sleep the Thread for a significant amount of time. This should definitely be done in a separate thread. The framework calls `TivoHandler#handleCommand` in a synchronous way so it should return immediately.